x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/indexing/AsyncTwoPhaseIndexer.java

```diff
@@ -158,7 +158,13 @@ public synchronized boolean maybeTriggerAsyncJob(long now) {
                         if (r) {
                             nextSearch(ActionListener.wrap(this::onSearchResponse, this::finishWithSearchFailure));
                         } else {
-                            finishAndSetState();
+                            // If we transition to `STOPPED` this means the previous state was `STOPPING`.
+                            // This is because we are guaranteed to originally been in an INDEXING state if `stop` is called.
+                            // Treat this like a `checkState` call that is transitioning to a STOPPED state.
+                            if (finishAndSetState() == IndexerState.STOPPED) {
```

Comment:
@jimczi @hendrikmuhs after more thought and discussion, this seems like the simpliest and least intrusive change. 

- If `finishAndSetState() == ABORTING`, that means `onAbort` was called internally, nothing else needs to be done and just exit
- If `finishAndSetState() == STARTED`, that means it transitioned from `INDEXING` and should just exit
- If `finishAndSetState() == STOPPED`, that means it transitioned from `STOPPING` and should treat it like a normal stop and call the `doSaveState` logic from `checkState`

The indexer should NEVER be in the `STARTED` state when this branch is called. If it is, that means the child class did something unsupported and it should allow `finishAndSetState()` to throw.

Reply:
I still fail to understand why this is an issue, the code path if the indexation is over is https://github.com/elastic/elasticsearch/pull/46162/files#diff-01a4ee37232ad8becca8990702ebb4f0L355
Why can't we do the same here, e.g.: 
````
onFinish(ActionListener.wrap(
                        r -> doSaveState(finishAndSetState(), position.get(), () -> {}),
                        e -> doSaveState(finishAndSetState(), position.get(), () -> {})));
````
?
IMO there is a bug currently because `onFinish` is never call in this branch and `finishAndSetState` can call onAbort or onStop before we call onFinish again. It looks like this disrepancy was added in https://github.com/elastic/elasticsearch/pull/45511  so we should ensure that we don't call onFinish if onAbort or onStop was already called.

Reply:
@jimczi you are right. All this special handling is due to a problem in our child class not in the `AsyncTwoPhaseIndexer`. I will change it to `onFinish` as that unifies the call flow.